src/credentials/sso/ssoAccessTokenProvider.ts

```diff
@@ -0,0 +1,148 @@
+    private ssoUrl: string
+    private ssoOidcClient: AWS.SSOOIDC
+    private cache: DiskCache
+
+    constructor(ssoRegion: string, ssoUrl: string, ssoOidcClient: AWS.SSOOIDC, cache: DiskCache) {
+        this.ssoRegion = ssoRegion
+        this.ssoOidcClient = ssoOidcClient
+        this.cache = cache
+        this.ssoUrl = ssoUrl
+    }
+
+    public async accessToken(): Promise<AccessToken> {
+        const accessToken = this.cache.loadAccessToken(this.ssoUrl)
+        if (accessToken) {
+            return accessToken
+        }
+
+        try {
+            const token = await this.pollForToken()
+            this.cache.saveAccessToken(this.ssoUrl, token)
+            return token
+        } catch (err) {
+            throw err
+        }
+    }
+
+    private async pollForToken(): Promise<AccessToken> {
+        const registration = await this.registerClient()
+        const authorization = await this.authorizeClient(registration)
+        const deviceCodeExpiration = Date.now() + authorization.expiresIn * 1000
+
+        //temporary logging to get URL
+        const logger = getLogger()
+        logger.info(
+            `To complete authentication for this SSO account, please continue to this SSO portal:${authorization.verificationUriComplete}`
+        )
+
+        // The retry interval converted to milliseconds
+        let retryInterval = authorization.interval * 1000
+
+        const createTokenParams = {
+            clientId: registration.clientId,
+            clientSecret: registration.clientSecret,
+            grantType: GRANT_TYPE,
+            deviceCode: authorization.deviceCode,
+        }
+
+        while (true) {
+            try {
+                const tokenResponse = await this.ssoOidcClient.createToken(createTokenParams).promise()
+                const accessToken: AccessToken = {
+                    startUrl: this.ssoUrl,
+                    region: this.ssoRegion,
+                    accessToken: tokenResponse.accessToken!,
+                    expiresAt: new Date(Date.now() + tokenResponse.expiresIn! * 1000).toISOString(),
+                }
+                return accessToken
+            } catch (err) {
+                if (err.code === 'SlowDownException') {
+                    retryInterval += 5000
+                } else if (err.code === 'AuthorizationPendingException') {
+                    if (Date.now() + retryInterval > deviceCodeExpiration) {
+                        throw Error(
+                            `Device code has expired while polling for SSO token, login flow must be re-initiated.`
+                        )
+                    }
+                    // else, wait the interval and try again
+                } else if (err.code === 'ExpiredTokenException') {
+                    throw Error(`Device code has expired while polling for SSO token, login flow must be re-initiated.`)
+                }
+            }
+            setTimeout(() => {}, retryInterval)
```

Comment:
Does this work? You're not awaiting the timeout. Worst case scenario, does this block other parts of the toolkit?

Reply:
This did not work. Thanks for your help creating a working solution.